### PR TITLE
ListKey sorting

### PR DESCRIPTION
**Pull request by Rishi-web**
W.r.t - https://github.ibm.com/kms/kmk/issues/1177
Supporting go sdk to list key in a sorted manner

### CODE REVIEW

**Comment by Rishi-web** [comment]
List key by `creationDate`  in descending order and `state` in ascending order

Screenshot 2022-06-01 at 5 09 21 PM

**Comment by Rishi-web** [comment]
go test 
Screenshot 2022-06-01 at 5 09 14 PM

File: kp.go [hunk 1]
```diff
@@ -508,3 +508,101 @@ func redact(s string, redactStrings []string) string {
...
+	*s = append(*s, "id")
+	return s
+}
+func (s *SortBy) WithIDDesc() *SortBy {
+	*s = append(*s, "-id")
+	return s
+}
+
+// sort by creation date
+func (s *SortBy) WithCretionDate() *SortBy {
```
Comment (mrodden) [inline kp.go:523]:
```suggestion
func (s *SortBy) WithCreationDate() *SortBy {
```

File: kp.go [hunk 2]
```diff
@@ -508,3 +508,101 @@ func redact(s string, redactStrings []string) string {
...
+	*s = append(*s, "creationDate")
+	return s
+}
+
+func (s *SortBy) WithCretionDateDesc() *SortBy {
```
Comment (mrodden) [inline kp.go:528]:
```suggestion
func (s *SortBy) WithCreationDateDesc() *SortBy {
```

File: README.md [hunk 3]
```diff
@@ -213,4 +213,42 @@ if err != nil {
...
\ No newline at end of file
+```
+
+### Fetching List Key In Sorted Ascending Order Based On Paramaeters.
+
+```go
+var sortObj SortBy
+sortStr := SortBy(sortObj)
+sortStr = sortStr.WithID()
+sortStr = sortStr.WithCretionDate()
```
Comment (mrodden) [inline README.md:224]:
```suggestion
sortStr = sortStr.WithCreationDate()
```

File: README.md [hunk 4]
```diff
@@ -213,4 +213,42 @@ if err != nil {
...
+    fmt.Println(err)
+}
+fmt.Println(keys)
+```
+
+### Fetching List Key In Sorted Descending Order Based On Paramaeters.
+
+```go
+var sortObj SortBy
+sortStr := SortBy(sortObj)
```
Comment (mrodden) [inline README.md:241]: Why did you make it an extended type vs doing a composition/encapsulation?

i.e. why not
```
type SortBy struct {
    sortStr string[]
}

...

mysort := SortBy{}.WithCreationDateDesc();

res := ListKeys(...., ListKeysOptions{sortby: mySort});

```

**Comment by Rishi-web** [comment]
List key with descending order by `id`
Screenshot 2022-07-04 at 5 34 10 PM

**Comment by Rishi-web** [comment]
List key in descending order by `creationDate` and ascending order of `imported`
Screenshot 2022-07-04 at 5 35 07 PM

**Comment by Rishi-web** [comment]
Test case result
Screenshot 2022-07-04 at 6 11 41 PM

File: README.md (hunk 4 above)
Comment (Rishi-web) [reply README.md]: I changed the design pattern to handle it differently, let me know does it achieve what you are suggesting

File: keys.go [hunk 5]
```diff
@@ -81,6 +81,7 @@ type Key struct {
 	PurgeAllowedFrom    *time.Time  `json:"purgeAllowedFrom,omitempty"`
 	PurgeScheduledOn    *time.Time  `json:"purgeScheduledOn,omitempty"`
 	DualAuthDelete      *DualAuth   `json:"dualAuthDelete,omitempty"`
+	Sort                string      `json:"sort,omitempty"`
```
Comment (mrodden) [inline keys.go:84]: I don't think this is needed or correct?

Its not used in the example tests either.

File: kp_test.go [hunk 6]
```diff
@@ -4145,3 +4145,199 @@ func TestSyncAssociatedResourcesError(t *testing.T) {
...
+
+	sortStr, _ := GetKeySortStr(WithCreationDate(), WithState())
+
+	listKeysOptions := &ListKeysOptions{
+		Sort: sortStr,
+	}
+	keys, err := c.ListKeys(context.Background(), listKeysOptions)
+	assert.NoError(t, err)
+	assert.NotNil(t, keys)
+
```
Comment (mrodden) [inline kp_test.go:4229]: Need to ensure that the correct HTTP call was actually made in both tests inside the function here

```suggestion
assert.True(t, gock.IsDone(), "Expected HTTP requests not called")
```

File: kp_test.go [hunk 7]
```diff
@@ -4145,3 +4145,199 @@ func TestSyncAssociatedResourcesError(t *testing.T) {
...
+	sortStr, _ = GetKeySortStr(WithLastRotateDateDesc())
+	listKeysOptions = &ListKeysOptions{
+		Sort: sortStr,
+	}
+
+	keys, err = c.ListKeys(context.Background(), listKeysOptions)
+	assert.NoError(t, err)
+	assert.NotNil(t, keys)
+
```
Comment (mrodden) [inline kp_test.go:4303]: Need to ensure that the correct HTTP call was actually made in both tests inside the function here

```suggestion
assert.True(t, gock.IsDone(), "Expected HTTP requests not called")
```
```

File: kp_test.go [hunk 8]
```diff
@@ -4145,3 +4145,199 @@ func TestSyncAssociatedResourcesError(t *testing.T) {
...
+	sort := "ids"
+	listKeysOptions = &ListKeysOptions{
+		Sort: &sort,
+	}
+
+	keys, err = c.ListKeys(context.Background(), listKeysOptions)
+	assert.Error(t, err)
+	assert.Nil(t, keys)
+	assert.Contains(t, err.Error(), "INVALID_QUERY_PARAM_ERR")
```
Comment (mrodden) [inline kp_test.go:4342]: Need to ensure that the correct HTTP call was actually made in both tests inside the function here
```suggestion
	assert.Contains(t, err.Error(), "INVALID_QUERY_PARAM_ERR")
	assert.True(t, gock.IsDone(), "Expected HTTP requests not called")
```

**Review by mrodden — CHANGES_REQUESTED** [review]

**Comment by mrodden** [comment]
Also it would be prudent to move the ListKey and Sort related functionality into a new file, but in the same package ("kp") to keep the kp.go file from getting too big / messy. Maybe something like `listkeys.go`

**Comment by Rishi-web** [comment]
Sort the keys based on params

Screenshot 2022-07-29 at 1 37 29 PM

**Comment by Rishi-web** [comment]
Screenshot 2022-07-29 at 1 38 26 PM

**Comment by Rishi-web** [comment]
Test case - sort keys
Screenshot 2022-07-29 at 1 38 53 PM

**Comment by Rishi-web** [comment]
Listing only standard keys
Screenshot 2022-07-29 at 1 46 03 PM

**Comment by Rishi-web** [comment]
Listing keys with `Limit = 1, offset = 0 and extractable = true`

Screenshot 2022-07-29 at 1 48 26 PM

**Comment by Rishi-web** [comment]
Filtering key based on state `Destroyed`

Screenshot 2022-07-29 at 1 51 29 PM

**Comment by Rishi-web** [comment]
Listing both `active and destroyed` state keys 

Screenshot 2022-07-29 at 1 59 05 PM

**Review by dinesh-venkatraman-g — CHANGES_REQUESTED** [review]
Can you move all the listkeys related entities and functions to the new file?

**Comment by dinesh-venkatraman-g** [comment]
Also, to avoid DCO failure, please follow this:
`To avoid having PRs blocked in the future, always include Signed-off-by: Author Name <authoremail@example.com> in every commit message. You can also do this automatically by using the -s flag (i.e., git commit -s).`

**Comment by Rishi-web** [comment]
Screenshot 2022-08-02 at 11 21 38 AM

Screenshot 2022-08-02 at 11 22 10 AM

**Review by dinesh-venkatraman-g — APPROVED** [review]

**Review by mrodden — APPROVED** [review]

**Pull request merged**